### PR TITLE
Adds flag_name to opts [ES-333]

### PR DESCRIPTION
**Pull request by jekku**
## What's here?

- We want to pass `flag_name` to coveralls.io via this library

## Milestones

- [x] Add flag_name to opts
- [x] Tests

## Usage samples

- Via Post

image

image

- Via Circle CI

image

image

- Via Circle CI **without flagname**

image

## References

- https://duffel.atlassian.net/browse/ES-333

### CODE REVIEW

**Comment by SamHutchings** [comment]
looks like we're only casting this to an empty string if nil in the `.post` task, which is used when you post to coveralls directly from the cli. 

what would happen if this wasn't provided in, say, the circleci task? is `flag_name: null` an acceptable value? it may well be, just want to double check. 

if you want to test uploading directly, you can enable one of your public github repos on coveralls.io and get the token, then run `coveralls.post` with the token value, like `MIX_ENV="test" mix cmd --app app_one mix coveralls.post -t {token} -n local -b pr-comment`

File: README.md [hunk 1]
```diff
@@ -169,6 +169,18 @@ Usage: mix coveralls.post <Options>
     -s (--sha)          Commit SHA (required when not using Travis)
     --build             Service number ('BUILDS' column at coveralls.io page)
     --parallel          coveralls.io 'parallel' option (See coveralls.io API Reference)
+    --flagname         coveralls.io 'flag_name' option (See coveralls.io API Reference)
```
Comment (SamHutchings) [inline README.md]: conflict marker here - also this bit of the readme is specifically talking about the `post` task, as this is common to all tasks we should have it in the section above (line 131)

File: lib/excoveralls/task/util.ex [hunk 2]
```diff
@@ -54,6 +54,17 @@ Usage: mix coveralls.post <Options>
     -s (--sha)          Commit SHA (required when not using Travis)
     --build             Service number ('BUILDS' column at coveralls.io page)
     --parallel          coveralls.io 'parallel' option (See coveralls.io API Reference)
+    --flagname         coveralls.io 'flag_name' option (See coveralls.io API Reference)
```
Comment (SamHutchings) [inline lib/excoveralls/task/util.ex]: same thing here, this is the `.post` section.

File: lib/mix/tasks.ex [hunk 3]
```diff
@@ -47,6 +48,15 @@ defmodule Mix.Tasks.Coveralls do
         all_options
       end
 
+    all_options =
```
Comment (SamHutchings) [inline lib/mix/tasks.ex]: this seems a bit unnecessary just to massage the option to the name we want. we could just do `flag_name: options[:flagname]` in the other tasks, right?

**Review by SamHutchings — COMMENTED** [review]
generally lgtm - couple of small points.

File: lib/mix/tasks.ex (hunk 3 above)
Comment (jekku) [reply lib/mix/tasks.ex]: good catch!

File: README.md [hunk 4]
```diff
@@ -169,6 +170,16 @@ Usage: mix coveralls.post <Options>
...
+    --subdir            Git repo sub directory: This will be added to the the front of file path, use if your covered
+                        file paths reside within a subfolder of the git repo. Example: If your source file path is
+                        "test.ex", and your git repo root is one directory up making the file's relative path
+                        "src/lib/test.ex", then the sub directory should be: "src/lib" (from coveralls.io)
+    --rootdir           This will be stripped from the file path in order to resolve the relative path of this repo's
+                        files. It should be the path to your git repo's root on your CI build environment. This is not
+                        needed if your source file path is already relative. It's used to pull the source file from the
+                        github repo, so must be exact. Example: If your source file path is "/home/runs/app/test.ex",
+                        and your git repo resides in "app", then the root path should be: "/home/runs/app/" (from
+                        coveralls.io)
```
Comment (ulissesalmeida) [inline README.md]: 🤔 where this docs came from? 

It seems unrelated

File: lib/excoveralls/task/util.ex [hunk 5]
```diff
@@ -54,6 +55,16 @@ Usage: mix coveralls.post <Options>
...
+    --subdir            Git repo sub directory: This will be added to the the front of file path, use if your covered
+                        file paths reside within a subfolder of the git repo. Example: If your source file path is
+                        "test.ex", and your git repo root is one directory up making the file's relative path
+                        "src/lib/test.ex", then the sub directory should be: "src/lib" (from coveralls.io)
+    --rootdir           This will be stripped from the file path in order to resolve the relative path of this repo's
+                        files. It should be the path to your git repo's root on your CI build environment. This is not
+                        needed if your source file path is already relative. It's used to pull the source file from the
+                        github repo, so must be exact. Example: If your source file path is "/home/runs/app/test.ex",
+                        and your git repo resides in "app", then the root path should be: "/home/runs/app/" (from
+                        coveralls.io)
```
Comment (ulissesalmeida) [inline lib/excoveralls/task/util.ex]: Same comment as above, I'm a little bit out of the loop, this doc seems unrelated with `flag_name`

When we were about to contribute back to the official library, is better have self contained commits to make the review and acceptance easier.

**Review by ulissesalmeida — COMMENTED** [review]
Looks pretty good for me, but I'm a little bit worried about the unrelated docs.

I don't know how we're organising here, but to contribute back the official library we'll need self contained commits.

File: README.md (hunk 1 above)
Comment (jekku) [reply README.md]: https://github.com/duffelhq/excoveralls/pull/1/commits/31b19c8d329de57954548710c9aa3d08b0e486b8

Fixed here :D

File: lib/excoveralls/task/util.ex (hunk 2 above)
Comment (jekku) [reply lib/excoveralls/task/util.ex]: https://github.com/duffelhq/excoveralls/pull/1/commits/31b19c8d329de57954548710c9aa3d08b0e486b8

Here as well

File: lib/mix/tasks.ex (hunk 3 above)
Comment (jekku) [reply lib/mix/tasks.ex]: https://github.com/duffelhq/excoveralls/pull/1/commits/a80c5e257a21593d06750580833e172b707bb0b3

Fixed on this end :)

File: README.md (hunk 4 above)
Comment (jekku) [reply README.md]: I think I botched the merge conflict resolution 😅 let m fix that

Comment (SamHutchings) [reply README.md]: these are the docs i moved from `post` to `common` in my earlier PR, i reckon this is a merge artifact. they can be removed. same for the next comment.

Comment (jekku) [reply README.md]: Fixed in: https://github.com/duffelhq/excoveralls/pull/1/commits/12d9d10c936b18e041f4337a1f539dcea645b4e1

File: lib/excoveralls/task/util.ex (hunk 5 above)
Comment (jekku) [reply lib/excoveralls/task/util.ex]: Fixed in: https://github.com/duffelhq/excoveralls/pull/1/commits/12d9d10c936b18e041f4337a1f539dcea645b4e1

**Comment by jekku** [comment]
Retest on latest build:

image

File: README.md [hunk 6]
```diff
@@ -169,7 +170,6 @@ Usage: mix coveralls.post <Options>
     -s (--sha)          Commit SHA (required when not using Travis)
     --build             Service number ('BUILDS' column at coveralls.io page)
     --parallel          coveralls.io 'parallel' option (See coveralls.io API Reference)
-```
```
Comment (SamHutchings) [inline README.md]: still need this to end the code block :)

File: lib/excoveralls/task/util.ex (hunk 5 above)
Comment (jekku) [reply lib/excoveralls/task/util.ex]: Wait still wrong

Comment (jekku) [reply lib/excoveralls/task/util.ex]: All good now :)

File: README.md (hunk 6 above)
Comment (jekku) [reply README.md]: Closed :) Looked at the diff and it seems clean

**Review by SamHutchings — APPROVED** [review]
lgtm!

**Comment by ulissesalmeida** [comment]
Nice!
